src/robot_state_publisher.cpp

```diff
@@ -68,110 +73,308 @@ geometry_msgs::msg::TransformStamped kdlToTransform(const KDL::Frame & k)
           child.getName().c_str());
       }
     } else {
       segments_.insert(make_pair(child.getJoint().getName(), s));
-      fprintf(stderr, "Adding moving segment from %s to %s\n", root.c_str(),
+      RCLCPP_DEBUG(get_logger(), "Adding moving segment from %s to %s", root.c_str(),
         child.getName().c_str());
     }
     addChildren(children[i]);
   }
 }
 
-
 // publish moving transforms
 void RobotStatePublisher::publishTransforms(
   const std::map<std::string, double> & joint_positions,
-  const builtin_interfaces::msg::Time & time,
-  const std::string & tf_prefix)
+  const builtin_interfaces::msg::Time & time)
 {
-  // fprintf(stderr, "Publishing transforms for moving joints\n");
+  RCLCPP_DEBUG(get_logger(), "Publishing transforms for moving joints");
   std::vector<geometry_msgs::msg::TransformStamped> tf_transforms;
 
   // loop over all joints
-  for (auto jnt = joint_positions.begin(); jnt != joint_positions.end(); ++jnt) {
-    auto seg = segments_.find(jnt->first);
+  for (const std::pair<std::string, double> & jnt : joint_positions) {
+    std::map<std::string, SegmentPair>::iterator seg = segments_.find(jnt.first);
     if (seg != segments_.end()) {
       geometry_msgs::msg::TransformStamped tf_transform =
-        kdlToTransform(seg->second.segment.pose(jnt->second));
+        kdlToTransform(seg->second.segment.pose(jnt.second));
       tf_transform.header.stamp = time;
-      tf_transform.header.frame_id = tf_prefix + "/" + seg->second.root;
-      tf_transform.child_frame_id = tf_prefix + "/" + seg->second.tip;
+      tf_transform.header.frame_id = seg->second.root;
+      tf_transform.child_frame_id = seg->second.tip;
       tf_transforms.push_back(tf_transform);
     }
   }
-  tf_broadcaster_.sendTransform(tf_transforms);
+  tf_broadcaster_->sendTransform(tf_transforms);
 }
 
 // publish fixed transforms
-void RobotStatePublisher::publishFixedTransforms(const std::string & tf_prefix, bool use_tf_static)
+void RobotStatePublisher::publishFixedTransforms()
 {
-  // fprintf(stderr, "Publishing transforms for fixed joints\n");
+  RCLCPP_DEBUG(get_logger(), "Publishing transforms for fixed joints");
   std::vector<geometry_msgs::msg::TransformStamped> tf_transforms;
   geometry_msgs::msg::TransformStamped tf_transform;
 
   // loop over all fixed segments
-  for (auto seg = segments_fixed_.begin(); seg != segments_fixed_.end(); seg++) {
-    geometry_msgs::msg::TransformStamped tf_transform = kdlToTransform(seg->second.segment.pose(0));
-    rclcpp::Time now = clock_->now();
-    if (!use_tf_static) {
-      now = now + rclcpp::Duration(std::chrono::milliseconds(500));  // 0.5sec in NS
+  for (const std::pair<std::string, SegmentPair> & seg : segments_fixed_) {
+    geometry_msgs::msg::TransformStamped tf_transform = kdlToTransform(seg.second.segment.pose(0));
+    rclcpp::Time now = this->now();
+    if (!use_tf_static_) {
+      now = now + rclcpp::Duration(std::chrono::milliseconds(500));
```

Comment:
Why are we adding 500ms to the current time here?

Reply:
I honestly have no idea.  This is ported directly from https://github.com/ros/robot_state_publisher/blob/fe661c1503cad7fedc8c16be6edddb5eb18d5a65/src/robot_state_publisher.cpp#L117 ; I'll have to look into it later.

Reply:
:ok_hand:

Reply:
I did a bit of archaeology in it, and found https://github.com/ros/robot_state_publisher/issues/46 .  In there, Jackie says that she considered this a hack, to be removed, but kept it for backwards-compatibility reasons.  It looks like she thought that we should just always default to use_tf_static, and remove the ability to shut it off.

That being said, I know we've had some complaints about this in the past, as splitting the tf tree between tf and tf_static can make it more difficult to get a full view of the tf tree.  So I'm on the fence about this.  My inclination is to leave it as-is, but I'm definitely open to hearing other opinions about it.

Reply:
I tend to agree with removing the `use_tf_static` option and always publishing fixed frames as static (i.e. `/tf_static`), on the grounds that the split topics `/tf` and `/tf_static` are agreed on as a standard. IMO, tools that interact with the TF package (e.g. viewing the tree) should handle data coming from both channels.

Moving forward, I'd suggest deprecating the usage of `use_tf_static`; defaulting to `true` and warning the user if they set it to `false`.